core/test/utils.hpp

```diff
@@ -32,154 +32,388 @@
+/**
+ * Merges two lists into a single list.
+ * The left and right list need to use the same type wrapper, which will also be
+ * the resulting wrapper containing elements of both lists. The order of the
+ * left and right list are preserved. The resulting list will have all elements
+ * of the left list, followed by all elements of the right list.
+ */
+template <typename FirstList, typename SecondList>
+struct merge_type_list {};
+
+template <template <typename...> class OuterWrapper, typename... Args1,
+          typename... Args2>
+struct merge_type_list<OuterWrapper<Args1...>, OuterWrapper<Args2...>> {
+    using type = OuterWrapper<Args1..., Args2...>;
+};
 
 
-using LocalGlobalIndexTypes =
-    ::testing::Types<std::tuple<int32, int32>, std::tuple<int32, int64>,
-                     std::tuple<int64, int64>>;
+/**
+ * This structure can change the outer type wrapper to the new, given one.
+ * Example:
+ * ```
+ * template <typename... Args>
+ * struct type_wrapper {};
+ * using old_list = std::tuple<int, double, short>;
+ * using new_list = typename change_outer_wrapper<type_wrapper, old_list>::type;
+ * // new_list = type_wrapper<int, double, short>;
+ * ```
+ */
+template <template <typename...> class NewOuterWrapper,
+          typename OldOuterWrapper>
+struct change_outer_wrapper {};
+
+template <template <typename...> class NewOuterWrapper,
+          template <typename...> class OldOuterWrapper, typename... Args>
+struct change_outer_wrapper<NewOuterWrapper, OldOuterWrapper<Args...>> {
+    using type = NewOuterWrapper<Args...>;
+};
 
 
-using PODTypes =
-#if GINKGO_DPCPP_SINGLE_MODE
-    ::testing::Types<float, gko::int32, gko::int64>;
-#else
-    ::testing::Types<float, double, gko::int32, gko::int64>;
-#endif
+/**
+ * Creates a type list (the outer wrapper stays the same) where each original
+ * type is wrapped into the given NewInnerWrapper. Example:
+ * ```
+ * using new_type =
+ *     typename add_internal_wrapper<std::complex,
+ *                                   std::tuple<float, double>>::type;
+ * // new_type = std::tuple<std::complex<float>, std::complex<double>>;
+ */
+template <template <typename...> class NewInnerWrapper, typename ListType>
+struct add_internal_wrapper {};
+
+template <template <typename...> class NewInnerWrapper,
+          template <typename...> class OuterWrapper, typename... Args>
+struct add_internal_wrapper<NewInnerWrapper, OuterWrapper<Args...>> {
+    using type = OuterWrapper<NewInnerWrapper<Args>...>;
+};
 
 
-using ValueAndIndexTypes =
-#if GINKGO_DPCPP_SINGLE_MODE
-    ::testing::Types<float, std::complex<float>, gko::int32, gko::int64,
-                     gko::size_type>;
-#else
-    ::testing::Types<float, double, std::complex<float>, std::complex<double>,
-                     gko::int32, gko::int64, gko::size_type>;
-#endif
+}  // namespace detail
+
+
+/**
+ * This type alias creates a cartesian product of the types in the left list
```

Comment:
there is a lot of duplicated documentation. Either here or in the `detail` namespace the documentation can be replaced by either `@copydoc` or just referencing the other doc

Reply:
I used the `@see` doxygen feature (but since we don't produce the documentation, I'm not 100% sure it works). I will try to compile it for myself to see if it works properly.
I also added details on how the implementation works in the `detail` namespace (mostly in `cartesian_type_product`).

Reply:
It is just shown as plain text (and not as a link). But the rest is displayed correctly.